Route creation reconciler loop.

cjcullen opened the pull request:
This teases apart node registration from cloudprovider routing configuration. It also makes route configuration robust against controller crashes and transient cloudprovider failures.

RIght now, I just have this naively syncing every 10 seconds. It would be nicer to react to changes in the NodeList, and then just run a background sync every few minutes. (@lavalamp, I'm told you're the expert on that :) )

I'd like to wait for #7984 to get in and then rebase this before merging.

cjcullen committed:
Route creation reconciler loop.

cjcullen commented on pkg/cloudprovider/gce/gce.go:
```diff
@@ -555,31 +560,45 @@ func getMetadataValue(metadata *compute.Metadata, key string) (string, bool) {
 	}
-	if err := gce.waitForGlobalOp(insertOp); err != nil {
-		if gapiErr, ok := err.(*googleapi.Error); ok && gapiErr.Code == http.StatusConflict {
-			// TODO (cjcullen): Make this actually check the route is correct.
```
The resolution of this TODO is that the reconciler will now remove an incorrect route and create a correct one. Any transient 409s or 404s from reconciling will be logged, but will not prevent progress.

davidopp commented on pkg/cloudprovider/routecontroller/routecontroller.go:
```diff
@@ -0,0 +1,145 @@
+		glog.Errorf("Ignoring route %s, unparsable CIDR: %v", route.Name, err)
+		return false
+	}
+	// Not responsible if this route's CIDR is not within our clusterCIDR
```
is -> if

cjcullen replied:
I think the comment is grammatically correct. Sounds like the wording might be a bit confusing. I'm open to suggestions.

davidopp replied:
Sorry, it actually makes total sense, I just mis-read it.

davidopp commented:
(Other than that, LGTM.)

cjcullen commented on pkg/cloudprovider/routecontroller/routecontroller_test.go:
```diff
@@ -0,0 +1,200 @@
+		if err := rc.reconcile(testCase.nodes, testCase.initialRoutes); err != nil {
+			t.Errorf("%d. Error from rc.reconcile(): %v", i, err)
+		}
+		time.Sleep(10 * time.Millisecond)
```
I feel gross about this. The reconciler kicks off goroutines to add/delete routes in the cloudprovider, so immediately checking the state of the fake_cloud in the main thread will show that nothing has been updated. A sleep of pretty much any length is enough to allow the background threads to update the in-memory representation, but it feels so wrong.

Any suggestions on a better way?

quinton-hoole replied:
Notify via a channnel that one is complete.

dchen1107 commented:
@cjcullen Have you run  e2e test with your change yet?

cjcullen commented:
Yes. All green (1 time). This was already rebased on top of #6949 though :/

cjcullen committed:
Revert "Revert "Modify nodes to register directly with the master.""

This reverts commit c53786ab31dc742708c449d6de9c4f170349f2a1.

cjcullen commented:
Rebased against #6949. Now running e2e's (over and over again).

cjcullen commented:
2 for 2 on e2es so far after the rebase (tearing down and rebuilding each time)

cjcullen committed:
Make routecontroller_test less hacky.
Rename reconcilePodCIDRs to reconcileNodeCIDRs.
Add comments and TODOs about using controller framework.

cjcullen commented:
This already got an LGTM before I rebased. Any opposition to me adding the status/LGTM tag?

lavalamp commented on pkg/cloudprovider/routecontroller/routecontroller.go:
```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		return fmt.Errorf("error listing routes: %v", err)
+	}
+	nodeList, err := rc.kubeClient.Nodes().List(labels.Everything(), fields.Everything())
```
// TODO: use pkg/controller/framework.NewInformer to watch this and reduce the number of lists needed.

cjcullen replied:
Done.

lavalamp commented on pkg/cloudprovider/nodecontroller/nodecontroller.go:
```diff
@@ -567,6 +340,9 @@ func (nc *NodeController) monitorNodeStatus() error {
 		return err
 	}
+	if nc.allocateNodeCIDRs {
+		nc.reconcileNodeCIDRs(nodes)
```
// TODO: use stuff from the controller framework to watch, to reduce the number of lists executed.

cjcullen replied:
Done.

lavalamp commented on pkg/cloudprovider/cloud.go:
```diff
@@ -80,12 +83,27 @@ type Instances interface {
+// Route is a representation of an advanced routing rule.
+type Route struct {
+	Name            string
+	TargetInstance  string
```
Comment on this-- is it the hostname, ip address, or name-as-given-by-cloud-provider?

cjcullen replied:
Done.

lavalamp commented on cluster/gce/util.sh:
```diff
@@ -772,7 +772,7 @@ function kube-down {
   local -a routes
   routes=( $(gcloud compute routes list --project "${PROJECT}" \
-              --regexp "${NODE_INSTANCE_PREFIX}-.+" | awk 'NR >= 2 { print $1 }') )
+    --regexp "${INSTANCE_PREFIX}-.{8}-.{4}-.{4}-.{4}-.{12}" | awk 'NR >= 2 { print $1 }') )
```
What does this line do? (please add a comment)

cjcullen replied:
Done.

lavalamp commented:
LGTM-- a few requests for comments

lavalamp commented:
Thank you very much, unfortunately I noticed one more thing. Sorry!

cjcullen commented on pkg/cloudprovider/routecontroller/routecontroller_test.go:
```diff
@@ -0,0 +1,212 @@
+		tick := time.Tick(10 * time.Millisecond)
+	poll:
+		for {
+			select {
```
@lavalamp is this more like what you were thinking?

lavalamp replied:
Yes, thank you, much better!

lavalamp commented on pkg/cloudprovider/routecontroller/routecontroller_test.go:
```diff
@@ -0,0 +1,212 @@
+		var finalRoutes []*cloudprovider.Route
+		var err error
+		timeoutChan := time.After(50 * time.Millisecond)
+		tick := time.Tick(10 * time.Millisecond)
```
I think you need to stop or delete this when you're done with it, though?

cjcullen replied:
Ahh. You're absolutely right. How's that?

lavalamp commented:
LGTM

dchen1107 commented:
LGTM. but still @cjcullen have you run e2e with this one yet?

cc/ @justinsb on aws, and @derekwaynecarr on vagrant. Just a FYI.

cjcullen commented:
e2e's were solid green 2 days ago. I'll rebase and run another round (or 3) to get a little more confidence that this will play nice with what we have at head.

dchen1107 commented:
@cjcullen Thanks for your patience.
